Takeon wrangler lambda lu 2539

sivagudur opened the pull request:
Please review my code

sivagudur committed:
Added Wrangler Lambda

sivagudur committed:
Added Unit Test

lcmorley commented on pom.xml:
```diff
@@ -0,0 +1,157 @@
+            <artifactId>aws-java-sdk-lambda</artifactId>
+            <version>1.11.560</version>
+        </dependency>
+
```
Is this extra line required?

lcmorley commented on pom.xml:
```diff
@@ -0,0 +1,157 @@
+            <artifactId>log4j-core</artifactId>
+            <version>2.11.2</version>
+        </dependency>
+        <dependency>
```
Perhaps for consistency a blank line should be put here.  Similarly for some of the other dependencies below.

sivagudur replied:
Sure I will leave a blank space after the end of each dependency

lcmorley commented on pom.xml:
```diff
@@ -0,0 +1,157 @@
+            </plugin>
+        </plugins>
+    </build>
+
```
One too many blank lines?

lcmorley commented on src/main/java/uk/gov/ons/validation/common/Constants.java:
```diff
@@ -0,0 +1,9 @@
+
+public final class Constants {
+
+        public static final String WRANGLER_NAME = "WRANGLER_NAME";
```
Formatting issues here. line 5 should be indented to the same level as the constructor below it.

lcmorley commented on src/main/java/uk/gov/ons/validation/entity/QuestionInputData.java:
```diff
@@ -0,0 +1,20 @@
+    private String questionCode;
+    private String response;
+
+
```
Too many blank lines.

lcmorley commented on src/main/java/uk/gov/ons/validation/entity/WranglerRequest.java:
```diff
@@ -0,0 +1,20 @@
+
+    private List<QuestionInputData> responses;
+
+
```
Too many blank lines.

lcmorley commented on src/main/java/uk/gov/ons/validation/entity/WranglerResponse.java:
```diff
@@ -0,0 +1,17 @@
+public class WranglerResponse {
+
+    private WranglerResponseData responses;
+}
```
Should have a blank line here.

lcmorley commented on src/main/java/uk/gov/ons/validation/entity/WranglerResponseData.java:
```diff
@@ -0,0 +1,20 @@
+@AllArgsConstructor
+@NoArgsConstructor
+public class WranglerResponseData {
+    public String primaryValue ;
```
Should have a blank line here and remove the excess lines at the end of this class.

lcmorley commented on src/main/java/uk/gov/ons/validation/serverless/WranglerLambdaHandler.java:
```diff
@@ -0,0 +1,36 @@
+            log.error("An exception was raised handling the Wrangler Lambda request.", e);
+            return "Failed";
+        }
+
```
Too many blank lines.

lcmorley commented on src/main/java/uk/gov/ons/validation/serverless/WranglerLambdaHandler.java:
```diff
@@ -0,0 +1,36 @@
+
+    private ProcessWranglerQuestionData newProcessQuestionData(){
+        return new ProcessWranglerQuestionData();
+
```
Is this blank line required?

sivagudur committed:
Added documentation in Readme.md

lcmorley commented on src/main/java/uk/gov/ons/validation/service/ProcessWranglerQuestionData.java:
```diff
@@ -0,0 +1,146 @@
+
+        log.info("Parsing Json Array to Java object: ");
+
+        //1. Building Validation Config this is separate feature. So hard coding it
```
I would say that as this is just hardcoded then it should be extracted and put into a separate class so that it doesn't make this class look too busy.

lcmorley commented on src/main/java/uk/gov/ons/validation/service/ProcessWranglerQuestionData.java:
```diff
@@ -0,0 +1,146 @@
+        List<QuestionInputData> responses = request.getResponses();
+
+        log.info("Request Data {} " , responses);
+
```
Too many blank lines.

lcmorley commented on src/main/java/uk/gov/ons/validation/service/ProcessWranglerQuestionData.java:
```diff
@@ -0,0 +1,146 @@
+        String finalDerivedQuestCode = null;
+        String finalQuestCodeValue = null;
+        String finalDerivedQuestValue = null;
+
```
Again, too many blank lines.

lcmorley commented on src/main/java/uk/gov/ons/validation/service/ProcessWranglerQuestionData.java:
```diff
@@ -0,0 +1,146 @@
+    private static final String MATCH_FOUND_DERIVED_QUEST = "Matching object found for Derived Question and its code is %s.";
+
+
+    public void processQuestionAndDerivedData(WranglerRequest request) throws Exception {
```
This method is very procedural at the moment. A method should really contain at the most 10-15 lines and have a specific function.

lcmorley commented on src/main/java/uk/gov/ons/validation/service/ProcessWranglerQuestionData.java:
```diff
@@ -0,0 +1,146 @@
+        log.info("Request Data {} " , responses);
+
+
+        String finalQuestCode = null;
```
Could this information be encapsulated in a class?

sivagudur committed:
Implemented Review Comments

sivagudur committed:
Formatted All Java Files

sivagudur commented:
Hi @lcmorley  I implemented the following review Comments. Can you please review. Thanks.

1. Created InvokeConfig and moved all logic
2. Removed blank lines in all files

lcmorley commented on src/main/java/uk/gov/ons/validation/entity/ValidationConfig.java:
```diff
@@ -0,0 +1,32 @@
+@NoArgsConstructor
+@AllArgsConstructor
+public class ValidationConfig {
+    public String questionCode;
```
Are these two attributes required?

sivagudur replied:
Yes required.

lcmorley commented on src/main/java/uk/gov/ons/validation/entity/WranglerResponse.java:
```diff
@@ -0,0 +1,16 @@
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class WranglerResponse {
```
Would this be a ValidationRequest rather than a WranglerResponse? A response is a reply to the caller whereas this is a request to the next lambda in line.

sivagudur replied:
Yes Renamed to ValidationRequest.  Done

lcmorley commented on src/main/java/uk/gov/ons/validation/service/InvokeConfig.java:
```diff
@@ -0,0 +1,57 @@
+     * @param inputData
+     * @param configQuestionCode
+     */
+    public void processQuestionCodeFound(QuestionInputData inputData, String configQuestionCode) {
```
I would have thought this method should just be called processQuestionCode.

sivagudur replied:
Done

lcmorley commented on src/main/java/uk/gov/ons/validation/service/InvokeConfig.java:
```diff
@@ -0,0 +1,57 @@
+    private String finalDerivedQuestCode;
+    private String finalQuestCodeValue;
+    private String finalDerivedQuestValue;
+    private boolean isQuestionCodeFound;
```
I don't think you need these extra boolean attributes to complicate things. You could have one boolean method which returns true if both the question and derived question information is found.

sivagudur replied:
Extra Boolean will give detail explanation whether Question Match found and in a similar way Derived Question Found. I added another method isBothFound() which verifies both Question and Derived question found.

lcmorley commented on src/main/java/uk/gov/ons/validation/service/ProcessWranglerQuestionData.java:
```diff
@@ -0,0 +1,118 @@
+                        log.info(format(QUES_DER_MESSAGE, invokeConfig.getFinalQuestCode(),
+                                invokeConfig.getFinalDerivedQuestCode(), invokeConfig.getFinalQuestCodeValue(),
+                                invokeConfig.getFinalDerivedQuestValue()));
+                        //Call External Lambda which performs Validation
```
You do not really need this comment. The method name alone should be ample to show we are calling a lambda.

sivagudur replied:
Additional comment removed

lcmorley commented on src/main/java/uk/gov/ons/validation/service/ProcessWranglerQuestionData.java:
```diff
@@ -0,0 +1,118 @@
+     */
+    private void callValidationLambda(InvokeConfig config) throws JsonProcessingException {
+        try {
+            //Call External Lambda which performs Validation
```
Again, I don't think you need this comment. The method name tells us what is happening.

sivagudur replied:
I removed it. This method has extra logic and for a new developer it would be easy to know at what point this is called. Java Comment will always highlight and understand the code for any one who looks into the code in the first instance. I feel writing an extra comment will always help.

lcmorley replied:
Generally if comments are required it means the code is not clear enough to understand in the first place. In this case the method name "callValidationLambda" states what is happening in the method so the comment is superfluous.

lcmorley commented on src/main/java/uk/gov/ons/validation/service/ProcessWranglerQuestionData.java:
```diff
@@ -0,0 +1,118 @@
+import static java.lang.String.format;
+
+@Log4j2
+public class ProcessWranglerQuestionData {
```
This class would ideally have a set of unit tests as it contains the main logic for your lambda

sivagudur replied:
I discussed with @dllmorgan yesterday and decided to extend unit test case next time.

lcmorley replied:
The problem is that "next time" never happens. Next time, you would have worked on something new and would have forgotten your thought process when you originally worked on this code. It seems the only testing at the moment is uploading it to AWS and checking that it still runs as expected. The point of unit tests is to assert that the code does what it should as early as possible. Unit tests should not be an extra, they should be as important if not more important than the code itself. If someone were to come along and make a change to your code how would they know if it still works as you originally intended without tests. Unit tests do not and should not take a long time to write. However I am aware that I am flogging a dead horse so will be happy to leave as is.

sivagudur committed:
Renamed processQuestionCode method

sivagudur committed:
Removed one additional comment

sivagudur committed:
Renamed a class to Validation Request

lcmorley commented on src/main/java/uk/gov/ons/validation/entity/ValidationConfig.java:
```diff
@@ -0,0 +1,32 @@
+@NoArgsConstructor
+@AllArgsConstructor
+public class ValidationConfig {
+    public String questionCode;
```
There attributes should really be marked private.

sivagudur replied:
Done Leon

sivagudur committed:
Renamed a class to Validation Request
